pkg/cli/upgradeassistant/cmd/migrate/migrate.go

```diff
@@ -0,0 +1,211 @@
+}
+
+// V130ToV140 migrates data from v1.3.0 to v1.4.0 with the following tasks:
+// 1. Add field `SharedServices` for all projects
+// 2. Add field `ProductName` in field `Services` for all envs
+// 3. Change the ServiceTemplateCounterName format
+func V130ToV140() error {
+	fmt.Println("Migrating data from 1.3.0 to 1.4.0")
+
+	allServices, err := mongodb.NewServiceColl().ListMaxRevisions(nil)
+	if err != nil {
+		return err
+	}
+	allProjects, err := templaterepo.NewProductColl().List()
+	if err != nil {
+		return err
+	}
+	allEnvs, err := mongodb.NewProductColl().List(&mongodb.ProductListOptions{ExcludeStatus: setting.ProductStatusDeleting})
+	if err != nil {
+		return err
+	}
+
+	var updatedProjects []*templatemodels.Product
+	var updatedEnvs []*models.Product
+
+	// service name is unique before current version
+	serviceMap := make(map[string]*models.Service)
+	for _, s := range allServices {
+		serviceMap[s.ServiceName] = s
+	}
+
+	// if items in serviceMap is less than allServices (which means that there are more than one services with same name),
+	// we should stop here since the logic below may cause unexpected effects.
+	if len(serviceMap) < len(allServices) {
+		fmt.Println("Migration skipped")
+		return nil
+	}
+
+	// add field `SharedServices` for all projects
+	for _, project := range allProjects {
+		var sharedServices []*templatemodels.ServiceInfo
+		services := project.AllServiceInfoMap()
+		for name := range services {
+			service := serviceMap[name]
+			if service == nil {
+				continue
+			}
+			if service.ProductName != project.ProductName {
+				sharedServices = append(sharedServices, &templatemodels.ServiceInfo{Name: name, Owner: service.ProductName})
+			}
+		}
+
+		if len(sharedServices) > 0 {
+			project.SharedServices = sharedServices
+			updatedProjects = append(updatedProjects, project)
+		}
+
+	}
+
+	// add field `ProductName` in field `Services` for all envs
+	for _, env := range allEnvs {
+		for _, group := range env.Services {
+			for _, s := range group {
+				service := serviceMap[s.ServiceName]
+				if service == nil {
+					continue
+				}
+				s.ProductName = service.ProductName
+			}
+		}
+
+		updatedEnvs = append(updatedEnvs, env)
+	}
+
+	if err = templaterepo.NewProductColl().UpdateAll(updatedProjects); err != nil {
+		fmt.Printf("Failed to upgrade projects, err: %s", err)
+		return err
+	}
+
+	if err = mongodb.NewProductColl().UpdateAll(updatedEnvs); err != nil {
+		fmt.Printf("Failed to upgrade envs, err: %s", err)
+		return err
+	}
+
+	if err = UpdateServiceCounter(allServices); err != nil {
+		fmt.Printf("Failed to upgrade counters, err: %s", err)
+		return err
+	}
+
+	fmt.Println("Migration finished")
+
+	return nil
+}
+
+// V140ToV130 rollbacks the changes from v1.4.0 to v1.3.0 with the following tasks:
+// 1. Remove field `SharedServices` for all projects
+// 2. Remove field `ProductName` in field `Services` for all envs
+// 3. Revert the ServiceTemplateCounterName format
+func V140ToV130() error {
```

Comment:
In this case the data can be reversed to v1.3.0, but in some cases, the data cannot be reverted.